src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala

```diff
@@ -143,18 +143,60 @@ trait TreeAndTypeAnalysis extends Debugging {
                 .filterNot(x => x.isSealed && x.isAbstractClass && !isPrimitiveValueClass(x))
             }
 
-            // enumerate only direct subclasses,
-            // subclasses of subclasses are enumerated in the next iteration
-            // and added to a new group
-            def groupChildren(wl: List[Symbol],
-                              acc: List[List[Type]]): List[List[Type]] = wl match {
-              case hd :: tl =>
-                val children = enumerateChildren(hd)
-                groupChildren(tl ++ children, acc :+ filterChildren(children))
-              case Nil      => acc
+            val baseTypesForType = {
+              val baseTypesForType = mutable.Map[Type, Set[Type]]()
+
+              val wl = mutable.Queue[Type](sym.tpe)
+              while (wl.nonEmpty) {
+                // enumerate only direct subclasses,
+                // subclasses of subclasses are enumerated in the next iteration
+                // and added to a new group
+                val tpe = wl.dequeue()
+                val children = enumerateChildren(tpe.typeSymbol)
+                val filtered = filterChildren(children)
+                wl ++= filtered
+                baseTypesForType += (tpe -> tpe.baseTypeSeq.toList.toSet)
```

Comment:
Comparing `Type` with `==` and `hashCode` is almost always a bad idea.

Perhaps you can instead use the base class symbols, rather than the base types, for this book keeping.